### ISSUE TITLE
Class method can't infer class' generic type

### ISSUE THREAD

**Issue by Karasuni** [Microsoft/TypeScript#15130]
Bug first reported at https://github.com/ReactiveX/rxjs/issues/2545. Please check out this issue for context.

**TypeScript Version:**  2.1.6 / 2.2.1

**Code**
[Playground link](https://www.typescriptlang.org/play/#src=class%20Observable%3CT%3E%20%7B%0D%0A%20%20%20%20subscribe%3CT%3E(next%3A%20(value%3A%20T)%20%3D%3E%20void)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20ObservableActual%3CT%3E%20%7B%0D%0A%20%20%20%20subscribe%3CR%3E(next%3A%20(value%3A%20R)%20%3D%3E%20void)%20%7B%0D%0A%20%20%20%20%20%20%20%20%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Aclass%20CustomType%20%7B%0D%0A%20%20%20%20id%3A%20string%0D%0A%7D%0D%0A%0D%0Aconst%20v1%20%3D%20new%20Observable%3CCustomType%3E()%0D%0Av1.subscribe((a)%20%3D%3E%20a.id)%3B%20%2F%2Fyou%20see%20compiler%20error%2C%20T%20is%20inferred%20to%20%7B%7D%0D%0A%0D%0Aconst%20v2%20%3D%20new%20ObservableActual%3CCustomType%3E()%0D%0Av2.subscribe((a)%20%3D%3E%20a.id)%3B%20%2F%2Fthis%20is%20actual%20shape%20of%20generic%20you%20suggested%20to%20have%20subscribe%20have%20generic%0D%0A%0D%0A%2F%2Fto%20make%20type%20works%20with%20your%20suggestion%2C%20you%20have%20to%20specify%20type%20for%20each%20subscribe%20calls%20%0D%0Av1.subscribe%3CCustomType%3E((a)%20%3D%3E%20a.id)%3B%0D%0Av2.subscribe%3CCustomType%3E((a)%20%3D%3E%20a.id)%3B)
```ts
class Observable<T> {
    subscribe<T>(next: (value: T) => void) {
        
    }
}

class ObservableActual<T> {
    subscribe<R>(next: (value: R) => void) {
        
    }
}

class CustomType {
    id: string
}

const v1 = new Observable<CustomType>()
v1.subscribe((a) => a.id); //you see compiler error, T is inferred to {}

const v2 = new ObservableActual<CustomType>()
v2.subscribe((a) => a.id); //this is actual shape of generic you suggested to have subscribe have generic

//to make type works with your suggestion, you have to specify type for each subscribe calls 
v1.subscribe<CustomType>((a) => a.id);
v2.subscribe<CustomType>((a) => a.id);
```

**Expected behavior:**
`subscribe` should be able to infer the type of it's class correctly

**Actual behavior:**
You get a `TS7006` implicit 'any' type error.

**Comment by aluanhaddad** [comment]
This happens because of how the classes are declared
Consider:
```ts
// define a class `Observable` having a type parameter `T`
class Observable<T> {
    // define a method `subscribe` having a type parameter `T`.
    subscribe<T>(next: (value: T) => void) {
    }
}
const  observable = new Observable<{ id: 1 }>();
observable.subscribe(x => console.log(x.id)); // error: property 'id' does not exist on type '{ }'.
```
`subscribe`'s `T` declaration shadows the declaration of `T` in the enclosing scope, the `Observable` class.
 Since we do not pass a type argument for `T` in the call to subscribe, the compiler will try to infer one from the argument `x => void`. Since there is no way to infer `T` from that, it defaults to inferring `{ }`, which is safe.

In the second class, the exact same thing is happening except that the override does not shadow the type parameter `T`, instead declaring a . However a type parameter `R`. `R` is not inferred for the same reason as `T` in the example above is not inferred, so it fall backs to  `{ }`.

In short, the first class should look like
```ts
// define a class `Observable` having a type parameter `T`
class Observable<T> {
    // define a method `subscribe` taking a function `next` that takes a `T` and returns `void`.
    // `T` is lexically scoped.
    subscribe(next: (value: T) => void) {
    }
}
```
and the second class should look like (exactly the same)
```ts
class ObservableActual<T>{
    subscribe(next: (value: T) => void) {
    }
}
```

**Comment by aluanhaddad** [comment]
[Playground Link](https://www.typescriptlang.org/play/index.html#src=%2F%2F%20define%20a%20class%20%60Observable%60%20having%20a%20type%20parameter%20%60T%60%0D%0Aclass%20Observable%3CT%3E%20%7B%0D%0A%20%20%20%20%2F%2F%20define%20a%20method%20%60subscribe%60%20having%20a%20its%20very%20own%20type%20parameter%20%60T%60%20and%20taking%20%0D%0A%20%20%20%20%2F%2F%20a%20function%20%60next%60%20which%20takes%20a%20value%20of%20that%20very%20own%20type%20%60T%60%20and%20returns%20%60void%60.%0D%0A%20%20%20%20subscribe(next%3A%20(value%3A%20T)%20%3D%3E%20void)%20%7B%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0Aconst%20observable%20%3D%20new%20Observable%3C%7B%20id%3A%20number%20%7D%3E()%3B%0D%0Aobservable.subscribe(x%20%3D%3E%20console.log(x.id))%3B%20%2F%2F%20error%3A%20property%20'id'%20does%20not%20exist%20on%20type%20%7B%20%7D.)
